google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,327 @@
+      this->sleeper_(backoff_policy->OnCompletion())
+          .then([this, retry_policy, backoff_policy](future<void> sleep_fu) {
+            sleep_fu.get();  // finish future for sake of finishing
+            this->mu_.lock();
+            this->impl_ = this->stream_factory_();
+            auto start_fu = this->impl_->Start();
+            this->mu_.unlock();
+            start_fu.then([this](future<bool> start_future) {
+              if (!start_future.get()) {
+                this->RetryLoop(retry_policy, backoff_policy, retry_type);
+              }
+              if (retry_type == RetryType.Start) {
+                return;
+              }
+              this->reinitializer_(std::move(this->impl_))
+                  .then([this, retry_policy, backoff_policy](
+                            future<StatusOr<
+                                std::unique_ptr<AsyncStreamingReadWriteRpc<
+                                    RequestType, ResponseType>>>>
+                                reinitialize_future) {
+                    auto reinitialize_response = reinitialize_future.get();
+                    if (!reinitialize_response.ok()) {
+                      this->RetryLoop(retry_policy, backoff_policy, retry_type);
+                    }
+                    std::lock_guard(this->retry_loop_mu_);
+                    this->mu_.lock();
+                    this->impl_ = std::move(*reinitialize_response);
+                    this->mu_.unlock();
+                    if (this->read_promise_.has_value()) {
+                      this->read_promise_.value().set_value(absl::optional());
+                      this->read_promise_ = absl::optional();
+                    }
+                    if (this->write_promise_.has_value()) {
+                      this->write_promise_.value().set_value(false);
+                      this->write_promise_ = absl::optional();
+                    }
+                    this->in_retry_loop_ = false;
+                  });
+            });
+          });
+    }
+    status_mu_.lock();
+    status_promise_.set_value(Status(StatusCode.kAborted, "Permanent error"));
+    status_mu_.unlock();
+    std::lock_guard(this->retry_loop_mu_);
+    this->retry_loop_mu_ = false;
+  }
+
+  std::unique_ptr<RetryPolicy const> const retry_policy_prototype_;
+  std::unique_ptr<BackoffPolicy const> const backoff_policy_prototype_;
+  AsyncSleeper sleeper_;
+  StreamFactory<ResponseType, RequestType> const stream_factory_;
+  StreamReinitializer<ResponseType, RequestType> const reinitializer_;
+
+  std::mutex mu_;
+
+  std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+      impl_;  // ABSL_GUARDED_BY(mu_)
+
+  std::mutex retry_loop_mu_;
+
+  bool in_retry_loop_;  // ABSL_GUARDED_BY(retry_loop_mu_)
+  absl::optional<promise<absl::optional<ResponseType>>>
+      read_promise_;  // ABSL_GUARDED_BY(retry_loop_mu_)
+  absl::optional<promise<bool>>
+      write_promise_;  // ABSL_GUARDED_BY(retry_loop_mu_)
```

Comment:
I think the communication mechanism here is the wrong direction. In particular, what you need is tracking of outstanding Read/Write calls, not communication when the retry finishes.

You should instead have:

```
// Start in retry loop because stream is not yet initialized.
bool in_retry_loop_ = true;  // ABSL_GUARDED_BY(mu_)
// Shutdown cannot be called until there are no outstanding reads or writes.
absl::optional<future<void>> outstanding_read_;  // ABSL_GUARDED_BY(mu_)
absl::optional<future<void>> outstanding_write_;  // ABSL_GUARDED_BY(mu_)
```

Then, you can initialize outstanding_read_ and outstanding_write_ to a new future when there is a downcall, and finish the promise and set the member to nullopt (if it is not already) when the Read/Write future completes.

Before calling Finish(), you can then 1) set in_retry_loop_ to true and 2) wait asynchronously for both outstanding futures to complete before calling Finish().

You can chain these pretty easily with:

```
make_ready_future().then([<bind state>]{
  std::lock_guard<std::mutex> l{mu_};
  if (outstanding_read_) return *std::move(outstanding_read_);
  return make_ready_future();
}).then([<bind state>]{
  std::lock_guard<std::mutex> l{mu_};
  if (outstanding_write_) return *std::move(outstanding_write_);
  return make_ready_future();
}).then([<bind state>]{
  std::lock_guard<std::mutex> l{mu_};
  return impl_.Finish();
}).then(<Do retry with finish result>)
```